### PR TITLE
Feature/4790 new exposure windows collection

### PR DESCRIPTION
**Pull request by 30mar**
## Description
The New Exposure Windows metric includes all Exposure Windows (from the Risk Calculation) that ENF has newly identified since the previous submission of usage data.
**NOTE: Unit testing will be included in another branch**
[TechSpec](https://github.com/corona-warn-app/cwa-app-tech-spec/blob/a8abcbf65c4df247107929536d3a712a11ed18b5/docs/spec/data-donation-client-ppa.md#new-exposure-windows)
## Link to Jira
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-5101

## Screenshots
No UI involved

### CODE REVIEW

File: src/xcode/ENA/ENA/Source/Models/ExposureWindowsMetadata/ExposureWindowsMetadata.swift [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+import Foundation
+
+struct ExposureWindowsMetadata: Codable {
+	
+	// MARK: - Init
+	
+	init(
+		newExposureWindowsQueue: [SubmittionExposureWindow],
+		reportedExposureWindowsQueue: [SubmittionExposureWindow]) {
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Models/ExposureWindowsMetadata/ExposureWindowsMetadata.swift:13]:
```suggestion
		reportedExposureWindowsQueue: [SubmittionExposureWindow]
	) {
```

File: src/xcode/ENA/ENA/Source/Models/ExposureWindowsMetadata/ExposureWindowsMetadata.swift [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+	var hash: String?
+	var date: Date
+	
+	init(exposureWindow: ExposureWindow, transmissionRiskLevel: Int, normalizedTime: Double, hash: String?, date: Date) {
+		self.exposureWindow = exposureWindow
+		self.transmissionRiskLevel = transmissionRiskLevel
+		self.normalizedTime = normalizedTime
+		self.hash = hash
+		self.date = date
+	}
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Models/ExposureWindowsMetadata/ExposureWindowsMetadata.swift:57]: please add marks

File: src/xcode/ENA/ENA/Source/Workers/Store/SecureStore.swift [hunk 3]
```diff
@@ -451,6 +452,13 @@ extension SecureStore: PrivacyPreservingProviding {
 		get { kvStore["keySubmissionMetadata"] as KeySubmissionMetadata? ?? nil }
 		set {
 			kvStore["keySubmissionMetadata"] = newValue
```
Comment (pascal-brause) [inline src/xcode/ENA/ENA/Source/Workers/Store/SecureStore.swift:454]: here ist now the `analyticsSubmitter?.triggerSubmitData()` missing?

**Review by pascal-brause — CHANGES_REQUESTED** [review]

File: src/xcode/ENA/ENA/Source/Models/ExposureWindowsMetadata/ExposureWindowsMetadataService.swift [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+			return
+		}
+		self.clearReportedExposureWindowsQueueIfNeeded(store: store)
+		
+		let mappedSubmittionExposureWindows: [SubmittionExposureWindow] = calculation.mappedExposureWindows.map {
+			SubmittionExposureWindow(
+				exposureWindow: $0.exposureWindow,
+				transmissionRiskLevel: $0.transmissionRiskLevel,
+				normalizedTime: $0.normalizedTime,
+				hash: self.convertToData($0.exposureWindow)?.sha256String(),
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Models/ExposureWindowsMetadata/ExposureWindowsMetadataService.swift:23]:
```suggestion
				hash: convertToData($0.exposureWindow)?.sha256String(),
```

File: src/xcode/ENA/ENA/Source/Models/ExposureWindowsMetadata/ExposureWindowsMetadataService.swift [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+class ExposureWindowsMetadataService {
+	
+	// MARK: - Internal
+
+	func collectExposureWindows(from riskCalculation: RiskCalculationProtocol, store: Store) {
+		guard let calculation = riskCalculation as? RiskCalculation else {
+			Log.debug("Instance of riskCalculation couldn't be casted into RiskCalculation type", log: .ppa)
+			return
+		}
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Models/ExposureWindowsMetadata/ExposureWindowsMetadataService.swift:15]: maybe we should think about extending the protocol so you don't need that typecast
Might be also helpful for testing?

File: src/xcode/ENA/ENA/Source/Models/ExposureWindowsMetadata/ExposureWindowsMetadataService.swift [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+			)
+		}
+	}
+	
+	// MARK: - Private
+
+	private func clearReportedExposureWindowsQueueIfNeeded(store: Store) {
+		if let nonExpiredWindows = store.exposureWindowsMetadata?.reportedExposureWindowsQueue.filter({
+			guard let day = Calendar.current.dateComponents([.day], from: $0.date, to: Date()).day else {
+				Log.debug("Exposure Window is removed from reportedExposureWindowsQueue as the date componant is nil", log: .ppa)
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Models/ExposureWindowsMetadata/ExposureWindowsMetadataService.swift:52]:
```suggestion
				Log.debug("Exposure Window is removed from reportedExposureWindowsQueue as the date component is nil", log: .ppa)
```

File: src/xcode/ENA/ENA/Source/Models/ExposureWindowsMetadata/ExposureWindowsMetadataService.swift [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+				return false
+			}
+			return day < 15
+		}) {
+			store.exposureWindowsMetadata?.reportedExposureWindowsQueue = nonExpiredWindows
+		}
+	}
+	
+	private func convertToData(_ window: ExposureWindow) -> Data? {
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Models/ExposureWindowsMetadata/ExposureWindowsMetadataService.swift:61]: If I get it right this method is only used to later create an sha256 String?
```suggestion
	private func convertToData(_ window: ExposureWindow) -> String? {
```
Better create the sha256 string complete here?

File: src/xcode/ENA/ENA/Source/Models/ExposureWindowsMetadata/ExposureWindowsMetadataService.swift (hunk 4 above)
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Models/ExposureWindowsMetadata/ExposureWindowsMetadataService.swift:23]: see my comment on convertToData

**Review by KaiTeuber — CHANGES_REQUESTED** [review]

File: src/xcode/ENA/ENA/Source/Workers/Store/SecureStore.swift (hunk 3 above)
Comment (30mar) [reply src/xcode/ENA/ENA/Source/Workers/Store/SecureStore.swift]: Yeah was removed by accident with the conflicts, i pushed again

**Comment by KaiTeuber** [comment]
Submittion vs Submission - I'll prefer to use Submission

File: src/xcode/ENA/ENA/Source/Models/ExposureWindowsMetadata/ExposureWindowsMetadataService.swift (hunk 5 above)
Comment (30mar) [reply src/xcode/ENA/ENA/Source/Models/ExposureWindowsMetadata/ExposureWindowsMetadataService.swift:15]: Yeah i agree, i will do that

File: src/xcode/ENA/ENA/Source/Services/PPAnalyticsSubmitter/PPAnalyticsSubmitter.swift [hunk 8]
```diff
@@ -39,63 +39,61 @@ final class PPAnalyticsSubmitter: PPAnalyticsSubmitting {
...
+			self.probabilityToSubmitExposureWindows = configuration.privacyPreservingAnalyticsParameters.common.probabilityToSubmitExposureWindows
+			
 			if let token = ppacToken {
 				// Submit analytics data with injected ppac token
 				self.submitData(with: token, completion: completion)
 			} else {
 				self.generatePPACAndSubmitData(completion: completion)
 			}
-			
-			self.hoursSinceTestResultToSubmitKeySubmissionMetadata = configuration.privacyPreservingAnalyticsParameters.common.hoursSinceTestResultToSubmitKeySubmissionMetadata
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Services/PPAnalyticsSubmitter/PPAnalyticsSubmitter.swift:93]: Please don't delete this. I need this for my task.

**Review by naveeddotio — CHANGES_REQUESTED** [review]

File: src/xcode/ENA/ENA/Source/Services/PPAnalyticsSubmitter/PPAnalyticsSubmitter.swift (hunk 8 above)
Comment (30mar) [reply src/xcode/ENA/ENA/Source/Services/PPAnalyticsSubmitter/PPAnalyticsSubmitter.swift:93]: sorry removed by accident with the merge conflicts, will return it

**Comment by KaiTeuber** [comment]
There is.a folder missing in Xcode
Bildschirmfoto 2021-02-19 um 10 57 13

File: src/xcode/ENA/ENA/Source/Models/ExposureWindowsMetadata/ExposureWindowsMetadataService.swift (hunk 5 above)
Comment (KaiTeuber) [reply src/xcode/ENA/ENA/Source/Models/ExposureWindowsMetadata/ExposureWindowsMetadataService.swift:15]: After extending the protocol you do not need the typecast / guard anymore.

File: src/xcode/ENA/ENA/Source/Models/ExposureWindowsMetadata/ExposureWindowsMetadataService.swift [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+		
+		do {
+			let encoder = JSONEncoder()
+			windowData = try encoder.encode(window)
+
+		} catch {
+			Log.error("ExposureWindow Encoding error", log: .ppa, error: error)
+		}
+		return windowData?.sha256String()
+	}
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Models/ExposureWindowsMetadata/ExposureWindowsMetadataService.swift:72]:
```suggestion
	private func generateSHA256(_ window: ExposureWindow) -> String? {
		let encoder = JSONEncoder()
		do {
			let windowData = try encoder.encode(window)
			return windowData.sha256String()
		} catch {
			Log.error("ExposureWindow Encoding error", log: .ppa, error: error)
		}
		return nil
	}
```

**Review by KaiTeuber — CHANGES_REQUESTED** [review]

File: src/xcode/ENA/ENA/Source/Services/PPAnalyticsSubmitter/PPAnalyticsSubmitter.swift [hunk 10]
```diff
@@ -240,32 +240,37 @@ final class PPAnalyticsSubmitter: PPAnalyticsSubmitting {
...
-			$0.newExposureWindows = newExposureWindows
-			*/
 			$0.userMetadata = userMetadata
 			$0.clientMetadata = clientMetadata
 			$0.userMetadata = userMetadata
 			
 			if shouldIncludeKeySubmissionMetadata {
 				$0.keySubmissionMetadataSet = keySubmissionMetadata
 			}
+			$0.userMetadata = userMetadata
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Services/PPAnalyticsSubmitter/PPAnalyticsSubmitter.swift:258]: Duplicate assignment

File: src/xcode/ENA/ENA/Source/Services/PPAnalyticsSubmitter/PPAnalyticsSubmitter.swift [hunk 11]
```diff
@@ -240,32 +240,37 @@ final class PPAnalyticsSubmitter: PPAnalyticsSubmitting {
...
+			$0.clientMetadata = clientMetadata
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Services/PPAnalyticsSubmitter/PPAnalyticsSubmitter.swift:259]: Duplicate assignment

**Review by KaiTeuber — APPROVED** [review]

**Review by naveeddotio — CHANGES_REQUESTED** [review]

File: src/xcode/ENA/ENA/Source/Models/Metadata/ExposureWindowsMetadata/ExposureWindowsMetadata.swift [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+	
+	// MARK: - Internal
+	
+	// Exposure Windows to be added to the next Submission
+	var newExposureWindowsQueue: [SubmissionExposureWindow]
+	
+	// Exposure Windows that was sent in pervious Submissions
+	var reportedExposureWindowsQueue: [SubmissionExposureWindow]
+	
+	// Date used to delete Records in reportedExposureWindowsQueue which are older than 15 days
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Models/Metadata/ExposureWindowsMetadata/ExposureWindowsMetadata.swift:41]: Missing variable declaration?

File: src/xcode/ENA/ENA/Source/Services/PPAnalyticsSubmitter/PPAnalyticsSubmitter.swift (hunk 11 above)
Comment (30mar) [reply src/xcode/ENA/ENA/Source/Services/PPAnalyticsSubmitter/PPAnalyticsSubmitter.swift]: yes, i fixed it, please check again

**Review by naveeddotio — CHANGES_REQUESTED** [review]

File: src/xcode/ENA/ENA/Source/Models/Metadata/ExposureWindowsMetadata/ExposureWindowsMetadata.swift (hunk 12 above)
Comment (30mar) [reply src/xcode/ENA/ENA/Source/Models/Metadata/ExposureWindowsMetadata/ExposureWindowsMetadata.swift:41]: What do you mean ? what is missing?

Comment (30mar) [reply src/xcode/ENA/ENA/Source/Models/Metadata/ExposureWindowsMetadata/ExposureWindowsMetadata.swift]: fixed

**Review by pascal-brause — APPROVED** [review]

**Review by naveeddotio — APPROVED** [review]

**Pull request merged**